Crash after reboot: `ClassCastException`

Lawnchair bug report 2023年1月13日 21:43:46
version: 12.1.0 (f0e3822) (12010005)
commit: f0e3822
build.brand: motorola
build.device: rtwo
build.display: T1TR33.4-15-6
build.fingerprint: motorola/rtwo_cn/rtwo:13/T1TR33.4-15-6/5caf4-e13604:user/release-keys
build.hardware: qcom
build.id: T1TR33.4-15-6
build.manufacturer: motorola
build.model: XT2301-5
build.product: rtwo_cn
build.type: user
version.codename: REL
version.incremental: 5caf4-e13604
version.release: 13
version.sdk_int: 33
display.density_dpi: 446
isRecentsEnabled: false

```
error: Uncaught exception

java.lang.RuntimeException: Unable to start activity ComponentInfo{app.lawnchair/app.lawnchair.LawnchairLauncher}: java.lang.ClassCastException: androidx.activity.ComponentActivity$e cannot be cast to com.android.launcher3.Launcher$NonConfigInstance
 at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3776)
 at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3924)
 at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:5933)
 at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:5817)
 at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:71)
 at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
 at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
 at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
 at android.app.ClientTransactionHandler.executeTransaction(ClientTransactionHandler.java:61)
 at android.app.ActivityThread.handleRelaunchActivityLocally(ActivityThread.java:5885)
 at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2407)
 at android.os.Handler.dispatchMessage(Handler.java:111)
 at android.os.Looper.loopOnce(Looper.java:238)
 at android.os.Looper.loop(Looper.java:357)
 at android.app.ActivityThread.main(ActivityThread.java:8122)
 at java.lang.reflect.Method.invoke(Native Method)
 at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
 at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:957)
Caused by: java.lang.ClassCastException: androidx.activity.ComponentActivity$e cannot be cast to com.android.launcher3.Launcher$NonConfigInstance
 at com.android.launcher3.Launcher.crossFadeWithPreviousAppearance(Unknown Source:4)
 at com.android.launcher3.Launcher.onCreate(Unknown Source:192)
 at com.android.launcher3.BaseQuickstepLauncher.onCreate(Unknown Source:0)
 at app.lawnchair.LawnchairLauncher.onCreate(Unknown Source:12)
 at android.app.Activity.performCreate(Activity.java:8387)
 at android.app.Activity.performCreate(Activity.java:8366)
 at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1384)
 at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3757)
 ... 17 more
```

--- a/src/com/android/launcher3/Launcher.java
+++ b/src/com/android/launcher3/Launcher.java
@@ -3114,7 +3114,12 @@ public class Launcher extends StatefulActivity<LauncherState> implements Launche
      * updates.
      */
     private void crossFadeWithPreviousAppearance() {
-        NonConfigInstance lastInstance = (NonConfigInstance) getLastNonConfigurationInstance();
+        final NonConfigInstance lastInstance;
+        if (getLastNonConfigurationInstance() instanceof NonConfigInstance) {
+            lastInstance = (NonConfigInstance) getLastNonConfigurationInstance();
+        } else {
+            lastInstance = null;
+        }
 
         if (lastInstance == null || lastInstance.snapshot == null) {
             return;

--- a/src/com/android/launcher3/Launcher.java
+++ b/src/com/android/launcher3/Launcher.java
@@ -1196,7 +1196,7 @@ public class Launcher extends StatefulActivity<LauncherState> implements Launche
         LauncherState[] stateValues = LauncherState.values();
         LauncherState state = stateValues[stateOrdinal];
 
-        NonConfigInstance lastInstance = (NonConfigInstance) getLastNonConfigurationInstance();
+        NonConfigInstance lastInstance = (NonConfigInstance) getLastCustomNonConfigurationInstance();
         boolean forceRestore = lastInstance != null
                 && (lastInstance.config.diff(mOldConfig) & CONFIG_UI_MODE) != 0;
         if (forceRestore || !state.shouldDisableRestore()) {
@@ -3114,7 +3114,7 @@ public class Launcher extends StatefulActivity<LauncherState> implements Launche
      * updates.
      */
     private void crossFadeWithPreviousAppearance() {
-        NonConfigInstance lastInstance = (NonConfigInstance) getLastNonConfigurationInstance();
+        NonConfigInstance lastInstance = (NonConfigInstance) getLastCustomNonConfigurationInstance();
 
         if (lastInstance == null || lastInstance.snapshot == null) {
             return;

--- a/src/com/android/launcher3/Launcher.java
+++ b/src/com/android/launcher3/Launcher.java
@@ -1196,12 +1196,7 @@ public class Launcher extends StatefulActivity<LauncherState> implements Launche
         LauncherState[] stateValues = LauncherState.values();
         LauncherState state = stateValues[stateOrdinal];
 
-        final NonConfigInstance lastInstance;
-        if (getLastNonConfigurationInstance() instanceof NonConfigInstance) {
-            lastInstance = (NonConfigInstance) getLastNonConfigurationInstance();
-        } else {
-            lastInstance = null;
-        }
+        NonConfigInstance lastInstance = (NonConfigInstance) getLastNonConfigurationInstance();
         boolean forceRestore = lastInstance != null
                 && (lastInstance.config.diff(mOldConfig) & CONFIG_UI_MODE) != 0;
         if (forceRestore || !state.shouldDisableRestore()) {

--- a/src/com/android/launcher3/Launcher.java
+++ b/src/com/android/launcher3/Launcher.java
@@ -1196,7 +1196,12 @@ public class Launcher extends StatefulActivity<LauncherState> implements Launche
         LauncherState[] stateValues = LauncherState.values();
         LauncherState state = stateValues[stateOrdinal];
 
-        NonConfigInstance lastInstance = (NonConfigInstance) getLastNonConfigurationInstance();
+        final NonConfigInstance lastInstance;
+        if (getLastNonConfigurationInstance() instanceof NonConfigInstance) {
+            lastInstance = (NonConfigInstance) getLastNonConfigurationInstance();
+        } else {
+            lastInstance = null;
+        }
         boolean forceRestore = lastInstance != null
                 && (lastInstance.config.diff(mOldConfig) & CONFIG_UI_MODE) != 0;
         if (forceRestore || !state.shouldDisableRestore()) {

--- a/src/com/android/launcher3/Launcher.java
+++ b/src/com/android/launcher3/Launcher.java
@@ -3114,12 +3114,7 @@ public class Launcher extends StatefulActivity<LauncherState> implements Launche
      * updates.
      */
     private void crossFadeWithPreviousAppearance() {
-        final NonConfigInstance lastInstance;
-        if (getLastNonConfigurationInstance() instanceof NonConfigInstance) {
-            lastInstance = (NonConfigInstance) getLastNonConfigurationInstance();
-        } else {
-            lastInstance = null;
-        }
+        NonConfigInstance lastInstance = (NonConfigInstance) getLastNonConfigurationInstance();
 
         if (lastInstance == null || lastInstance.snapshot == null) {
             return;

